### PR TITLE
[air] Pass on KMS-related kwargs for s3fs

### LINKED ISSUE #35519: [Train] Syncer fails to write to S3 after 2 days of training
### What happened + What you expected to happen

training job was running for 2+ days using AWS VM but then  it ran into this error:
ERROR trainable.py:597 -- Could not upload checkpoint even after 3 retries.Please check if the credentials expired and that the remote filesystem is supported.. For large checkpoints, consider increasing `SyncConfig(sync_timeout)` (current value: 1800 seconds). Checkpoint URI: s3://a9vs-photon-fsx-data-repository/sync-with-fsx/Results/parisexp/TorchTrainer_2023-05-15_22-41-59/TorchTrainer_b3daa_00000_0_2023-05-15_22-42-00/checkpoint_000002
The job did not indicate logs from ray for some time with stuff like trial name, status and stuff after encountering the error but then recovered.  The VM is using IAM role which still has good s3 permissions. My syncconfig in TorchTrainer looks like 

SyncConfig(upload_dir="s3://a9vs-photon-fsx-data-repository/sync-with-fsx/Results/parisexp")

if pyarrow.fs.FileSystem.from_uri(uri) in remote_storage.py does not refresh credentials after a while, then not purging the cache after some time may be a bug in ray:
    cache_key = (parsed.scheme, parsed.netloc)

    if cache_key in _cached_fs:
        fs = _cached_fs[cache_key]
        return fs, path

    try:
        fs, path = pyarrow.fs.FileSystem.from_uri(uri)
        _cached_fs[cache_key] = fs
        return fs, path

### Versions / Dependencies

ray 2.3.1

### Reproduction script

use an s3 uri based syncconfig in ray Trainer and let the training run for 2 + days. 

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

### LINKED ISSUE #35561: [air / checkpoints] - Support KMS key encrypted S3 bucket for saving checkpoints
### Description

Ray's `checkpoint.Checkpoint.to_uri()` as described [here](https://docs.ray.io/en/latest/ray-air/api/doc/ray.air.checkpoint.Checkpoint.to_uri.html) allows users to write model checkpoints to a cloud storage eg. S3. Users should be given the option of passing in KMS arguments in cases when the S3 bucket being used for storage is KMS encrypted. 

### Use case

We are using an S3 bucket that is KMS key encrypted as our cloud storage. As a result our boto3 upload logic looks something like this (actual values have been replaced by placeholders).
```
ADHOC_BUCKET = "bucket-name"
_ADHOC_BUCKET_KMS_KEY_ID = "kms_key"
_STAGING_CANONICAL_ID = "staging_id"
_PROD_CANONICAL_ID = "prod_id"
_ACL_LIST = "id={},id={}".format(_PROD_CANONICAL_ID, _STAGING_CANONICAL_ID)
_ADHOC_PUT_REQUEST_EXTRA_ARGS = {
    "GrantFullControl": _ACL_LIST,
    "SSEKMSKeyId": _ADHOC_BUCKET_KMS_KEY_ID,
    "ServerSideEncryption": "aws:kms",
}

s3_client = boto3.client('s3', region_name="us-west-2", config=Config(signature_version="s3v4"))
response = s3_client.upload_file("file.txt", ADHOC_BUCKET, "common_assets/file.txt", ExtraArgs=_ADHOC_PUT_REQUEST_EXTRA_ARGS)
print(response)
```

Passing the ExtraArgs to `checkpoint.Checkpoint.to_uri()` is currently not possible as a result of which we are unable to save periodic checkpoints during model training.

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

We currently only parse and pass limited selection of options to s3fs. One recent request was related to passing KMS settings. This PR extends the s3 uri string to allow configuration of signature version, sse, sse key ID, and ACLs in s3 URIs if s3fs is used.

This PR also changes the fs caching logic, which is a requirement for options to be parsed again, e.g. when a key ID is changed in subsequent calls. FS cache keys now include the query string, and cache items are stale after 5 minutes and re-created. As a side-effect, this should fix any problems that come with cached filesystems, e.g. expiring credentials.

## Related issue number

Closes #35561
Might close #35519 (pending investigation)

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/air/_internal/remote_storage.py [hunk 1]
```diff
@@ -199,22 +207,38 @@ def _translate_s3_options(options: Dict[str, List[str]]) -> Dict[str, Any]:
...
+    option_map = {
+        "signature_version": "signature_version",
+    }
+    config_kwargs = _translate_options(option_map, options)
+
+    # s3_additional_kwargs
+    option_map = {
+        "ServerSideEncryption": "ServerSideEncryption",
+        "SSEKMSKeyId": "SSEKMSKeyId",
+        "GrantFullControl": "GrantFullControl",
```
Comment (krfricke) [inline python/ray/air/_internal/remote_storage.py:229]: It's hard to come up with a dynamic approach to cover all possible options here without churning the APIs too much. I'd like to revisit this when we think about custom upload logic and Syncer deprecation, and for now go with manually selected properties to forward.

Comment (justinvyu) [reply python/ray/air/_internal/remote_storage.py:229]: Yeah, I agree that we should revisit this. I was searching for other possibilities and `ACL` is another param that can be passed into `s3_additional_kwargs`. Then, GCS also has a bunch of query params that users can specify.

Another thing to note: with the current `Syncer` abstraction, users could technically just create an s3fs client and customize all of these settings there.

**Review by justinvyu — APPROVED** [review]
Looks good to me!

I do have some questions about getting rid of the cache:
- Should we calculate some before/after timing of the benefit this cache actually had?
- Regarding getting rid of the cache being a "requirement for options to be parsed again": how do users actually pass in these options again? They only get to specify `storage_path` once at the beginning, so even if the `KeyID` or `access_key` get refreshed, the query params would still be the same as before, and we'd run into the same problem. Am I missing something here?

**Comment by krfricke** [comment]
> Looks good to me!
> 
> I do have some questions about getting rid of the cache:
> 
>     * Should we calculate some before/after timing of the benefit this cache actually had?

I'll do some small-scale benchmarking before merge, but I don't expect any significant drawdowns as we call `get_fs_and_path` relatively rarely.

> 
>     * Regarding getting rid of the cache being a "requirement for options to be parsed again": how do users actually pass in these options again? They only get to specify `storage_path` once at the beginning, so even if the `KeyID` or `access_key` get refreshed, the query params would still be the same as before, and we'd run into the same problem. Am I missing something here?

It's mostly for calls to e.g. `Checkpoint.to_uri()` where the URI can be re-specified.

**Comment by krfricke** [comment]
Before:

```
> time python -c "from ray.air._internal.remote_storage import get_fs_and_path; [get_fs_and_path('s3://some') for i in range(10000)]"
python -c   1.21s user 2.53s system 428% cpu 0.872 total
```

After:
```
> time python -c "from ray.air._internal.remote_storage import get_fs_and_path; [get_fs_and_path('s3://some') for i in range(10000)]"
python -c   1.23s user 2.56s system 403% cpu 0.940 total
```

Looks like no significant speed reduction

**Comment by krfricke** [comment]
Hm, some tests are failing because we re-create the mock filesystem. It looks like we need to at least cache that.

A few options:
1. We continue to cache and include the query parameters in the caching key
2. We only cache the mock/memory filesystem
3. We patch get_fs_and_path for mock/memory tests
4. We use moto/s3 instead of mock/memory filesystems

1) would require to use a time-based cache to address #35519
2) means we would keep test-specific code in the library, which is not ideal
3) means boilerplate for the tests and mocking
4) would increase runtime of the tests

From these options I believe 1 is the best. We can keep the timeout relatively small (say a minute or so). 

I'll update the PR.

File: python/ray/air/_internal/remote_storage.py (hunk 1 above)
Comment (justinvyu) [reply python/ray/air/_internal/remote_storage.py:256]: Oh, I see, the custom `Syncer` doesn't affect `Checkpoint.to_uri` at all.

**Review by justinvyu — APPROVED** [review]
Ah I see, so `mock://` is something provided by pyarrow, and it gets wiped if we don't cache if for tests.

**Pull request merged**